Parameterize the game object, migrate from fixtures to factories, increase test coverage ahead of eCTF

rbclark opened the pull request:
Potentially still to do/check:

- [x] Verify when logged in as admin creating a game is possible
- [x] Add better helper text on game creation for admins

Bialogs commented on app/views/layouts/application.html.haml:
```diff
@@ -13,25 +13,26 @@
+                  %li{:class => "#{'active' if current_page? game_summary_path}"}
+                    %a{:href => game_summary_path} Summary
+                %li
+                  %a{:href => "http://mitrecyberacademy.org/contact", :target => "_blank"}
```
`"http://mitrecyberacademy.org/contact"` should be parameter

Bialogs commented on app/views/user_mailer/ranking.html.haml:
```diff
@@ -2,21 +2,19 @@
   Hello #{@user.full_name}!
 %p
   Congratulations on completing the MITRE CTF! Your team, #{@user.team.team_name} came ranked #{@rank.ordinalize}.
```
MITRE CTF!

Bialogs commented on app/views/user_mailer/ranking.html.haml:
```diff
@@ -2,21 +2,19 @@
+      #{@game.recruitment_additional_text}
 %p
   Thanks for participating!
 %p -The MITRE CTF Team
```
-The MITRE CTF Team should be parameter

rbclark commented on app/models/solved_challenge.rb:
```diff
@@ -11,11 +11,11 @@ class SolvedChallenge < FeedItem
 
   def description
-    %(Solved challenge "#{challenge.category.name} #{challenge.point_value}")
+    %(#{super.titleize} "#{challenge.category.name} #{challenge.point_value}")
```
Update, it ends up looking like:
```
pry> challenge.description
=> "Solved Challenge \"Web 100\""
```
which is acceptable since it will show in the UI as `Solved Challenge "Web 100"`

rbclark commented on test/mailers/user_mailer_test.rb:
```diff
@@ -10,135 +10,173 @@ def default_url_options
+    assert_equal [@first_place_user.email], email.to
+    assert_equal "#{@game.title}: Congratulations!", email.subject
     assert_equal true, email.has_attachments?
-    assert_includes (strip_tags(email.body.parts.first.to_s).squish.to_s), @rank_email_body
```
Can we still assert some basic checks here such as 'Your team came ranked x'? Otherwise having tests for every email use case is pretty useless since we can't verify that ranking is working correctly based on the known ranks in the tests and breaking variable interpolation in the emails would not raise any errors.

rbclark commented:
In setting up a new game, there are some slightly confusing options:

1. The scholarships_available checkbox enables the prizes_text textbox. It would be much better if we renamed prizes_text to scholarship_text. I had to look in the DB to verify this.
1. We should probably add some custom JS to make the fields hide and show.
1. The participant_recruitment_url has lost its purpose, it should probably be removed and replaced with the recruitment_text box since links should be able to put in the recruitment_text box. Currently participant_recruitment_url is not inserted anywhere.

rbclark commented:
Footer is also missing, and terms of service are not rendered anywhere even though they can but saved to the DB.

Bialogs commented:
Need to update the employment information on the user's team page.

Bialogs commented:
Need to have the ability to update the external mailer configurations (iremove hardcoded sendgrid), might do something like the admin rake task where you can log in and update the ActionMailer settings. Maybe too technical, but it gets it working for now.

Bialogs commented:
Need to be able to configure reCAPTCHA information the same way as mailer info.

Bialogs commented:
Closes #369

Bialogs commented:
Closes #75

Bialogs commented:
closes  #73

rbclark commented on app/models/team.rb:
```diff
@@ -136,36 +119,41 @@ def score
+
+  # Number of users on team is calculated using game team size
+  def set_slots_available
+    updated_slots_available = division.game.team_size - users.count
```
Saying `Game.instance` here is probably less error-prone and will use the cache. This will stop working if we ever add support for multiple games though.

rbclark commented on config/initializers/recaptcha.rb:
```diff
@@ -1,5 +0,0 @@
-Recaptcha.configure do |config|
-  config.site_key   = ENV['RECAPTCHA_SITE_KEY'] ||  '6LeIxAcTAAAAAJcZVRqyHh71UMIEGNQ_MXjiZKhI'
```
Removing this file will break Heroku badly, it needs to stay in the repository. Otherwise, on every deploy Heroku will delete this file.

rbclark commented on config/environments/production.rb:
```diff
@@ -96,14 +96,4 @@
 
   config.action_mailer.default_url_options = { host: ENV['APP_URL'] }
-
-  ActionMailer::Base.smtp_settings = {
```
We need to figure out a way to parameterize this that doesn't include removing it completely, perhaps we can get a default that is just all of these things as ENV's, that way this still works on Heroku.
